tensorflow/lite/micro/kernels/cmsis-nn/conv.cc

```diff
@@ -240,126 +239,119 @@ TfLiteStatus EvalQuantizedPerChannel(
+
+    // Initialize cmsis-nn per channel quantization parameters
+    cmsis_nn_per_channel_quant_params quant_params;
+    quant_params.multiplier =
+        const_cast<int32_t*>(data.per_channel_output_multiplier);
+    quant_params.shift = const_cast<int32_t*>(data.per_channel_output_shift);
+
+    RuntimeShape filter_shape = tflite::micro::GetTensorShape(filter);
+    RuntimeShape input_shape = tflite::micro::GetTensorShape(input);
+    RuntimeShape output_shape = tflite::micro::GetTensorShape(output);
+    RuntimeShape bias_shape = tflite::micro::GetTensorShape(bias);
+
+    // Consistency check.
+    TFLITE_DCHECK_LE(conv_params.activation.min, conv_params.activation.max);
+    TFLITE_DCHECK_EQ(input_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(filter_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(output_shape.DimensionsCount(), 4);
+    const int batch_size = MatchingDim(input_shape, 0, output_shape, 0);
+    const int input_depth = MatchingDim(input_shape, 3, filter_shape, 3);
+    const int output_depth = MatchingDim(filter_shape, 0, output_shape, 3);
+    if (tflite::micro::GetTensorData<int8_t>(bias)) {
+      TFLITE_DCHECK_EQ(bias_shape.FlatSize(), output_depth);
+    }
+
+    // Initialize cmsis-nn dimensions
+    // Input
+    cmsis_nn_dims input_dims;
+    input_dims.n = batch_size;
+    input_dims.h = input_shape.Dims(1);
+    input_dims.w = input_shape.Dims(2);
+    input_dims.c = input_depth;
+
+    // Filter
+    cmsis_nn_dims filter_dims;
+    filter_dims.n = output_depth;
+    filter_dims.h = filter_shape.Dims(1);
+    filter_dims.w = filter_shape.Dims(2);
+    filter_dims.c = input_depth;
+
+    // Bias
+    cmsis_nn_dims bias_dims;
+    bias_dims.n = 1;
+    bias_dims.h = 1;
+    bias_dims.w = 1;
+    bias_dims.c = output_depth;
+
+    // Output
+    cmsis_nn_dims output_dims;
+    output_dims.n = batch_size;
+    output_dims.h = output_shape.Dims(1);
+    output_dims.w = output_shape.Dims(2);
+    output_dims.c = output_depth;
+
+    // Initialize cmsis-nn context
+    cmsis_nn_context ctx;
+    ctx.buf = nullptr;
+    ctx.size = 0;
+
+    if (data.buffer_idx > -1) {
+      ctx.buf = context->GetScratchBuffer(context, data.buffer_idx);
+      // Note: ctx.size is currently not used in cmsis-nn.
+      // The buffer should be allocated in the Prepare function through
+      // arm_convolve_wrapper_s8_get_buffer_size
+    }
+
+    // arm_convolve_wrapper_s8 dispatches the optimized kernel accordingly with
+    // the parameters passed
+    TFLITE_DCHECK_EQ(
+        arm_convolve_wrapper_s8(
+            &ctx, &conv_params, &quant_params, &input_dims,
+            tflite::micro::GetTensorData<int8_t>(input), &filter_dims,
+            tflite::micro::GetTensorData<int8_t>(filter), &bias_dims,
+            tflite::micro::GetTensorData<int32_t>(bias), &output_dims,
+            tflite::micro::GetTensorData<int8_t>(output)),
+        ARM_MATH_SUCCESS);
   } else {
```

Comment:
Removing the macro will slightly increase the mem footprint for this CMSIS-NN kernel. But since dilation support will be available soon, this is only a temporary issue.

Reply:
can we make a github issue for supporting dilation in the cmsis kernels? and add a TODO(#link to issue): remove calls to reference kernels once dilation is supported with optimized code.